src/credentials/FabricTable.cpp

```diff
@@ -360,6 +360,53 @@ CHIP_ERROR FabricInfo::SetExternallyOwnedOperationalKeypair(P256Keypair * keyPai
     return CHIP_NO_ERROR;
 }
 
+CHIP_ERROR FabricInfo::ValidateNewNOCCredentials()
+{
+    P256PublicKey pubkey;
+    ValidationContext validContext;
+    validContext.Reset();
+    validContext.mRequiredKeyUsages.Set(KeyUsageFlags::kDigitalSignature);
+    validContext.mRequiredKeyPurposes.Set(KeyPurposeFlags::kServerAuth);
+
+    // Make sure to not modify any of our state until VerifyCredentials passes.
+    PeerId operationalId;
+    FabricId fabricId;
+    ChipLogProgress(Discovery, "Verifying the received credentials");
+    CHIP_ERROR err = VerifyCredentials(mNOCCert, mICACert, mRootCert, validContext, operationalId, fabricId, pubkey);
+    if (err != CHIP_NO_ERROR && err != CHIP_ERROR_WRONG_NODE_ID)
+    {
+        err = CHIP_ERROR_UNSUPPORTED_CERT_FORMAT;
+    }
+    ReturnErrorOnFailure(err);
+
+    VerifyOrReturnError(mOperationalKey != nullptr, CHIP_ERROR_INCORRECT_STATE);
+
+    // Verify that public key in NOC matches public key generated by node and sent in CSRResponse message.
+    VerifyOrReturnError(mOperationalKey->Pubkey().Length() == pubkey.Length(), CHIP_ERROR_INVALID_PUBLIC_KEY);
+    VerifyOrReturnError(memcmp(mOperationalKey->Pubkey().ConstBytes(), pubkey.Bytes(), pubkey.Length()) == 0,
+                        CHIP_ERROR_INVALID_PUBLIC_KEY);
+
+    if (mOperationalId == PeerId())
+    {
+        mOperationalId = operationalId;
+    }
+    else
+    {
+        VerifyOrReturnError(mOperationalId == operationalId, CHIP_ERROR_INVALID_ARGUMENT);
+    }
+
+    if (mFabricId == kUndefinedFabricId)
+    {
+        mFabricId = fabricId;
+    }
+    else
+    {
+        VerifyOrReturnError(mFabricId == fabricId, CHIP_ERROR_INVALID_ARGUMENT);
```

Comment:
That's not the error the spec says to send back here (for UpdateNOC).

Reply:
changed that to CHIP_ERROR_INSUFFICIENT_PRIVILEGE according to the spec